File: src/Assets/Modules/Trunk/Dialog/UI/DialogPanel.cs
using GameLib.DI;
using QS.Common;
using QS.Common.Util;
using QS.Player;
using QS.PlayerControl;
using System;
using System.Collections;
using System.Linq;
using System.Threading.Tasks;
using UnityEngine.Events;
using UnityEngine.UIElements;

namespace QS.UI
{
    public class OptionSelectedEvent : UnityEvent<int> { }

    /// <summary>
    /// UI也应当由对应的模块实现，这样子才能够保证高度内聚
    /// </summary>
    //[Scope(Lazy = false)]
    [Obsolete]
    public class DialogPanel : BaseDocument<DialogPanel.Props, DialogPanel.States>
    {
        [Injected]
        readonly PlayerControls playerActions;
        [Injected]
        readonly ITimer timer;

        public record Props
        {
            public string speaker;

            public Props(string speaker, string[] content)
            {
                this.speaker = speaker;
                this.lines = content;
            }

            public string[] lines;
        }

        public record States
        {
            public string displayedText = string.Empty;
            public int currentLine = 0;
        }

        public DialogPanel()
        {
            DefaultProps = new Props(string.Empty, new string[0]);
        }

        Label contentLabel;
        Label speakerLabel;

        const float c_PrintInterval = 0.5f;

        public UnityEvent OnDisplayFullCotent = new();
        public OptionSelectedEvent OnOptionSelected = new();

        protected override Task OnDocumentLoaded()
        {
            Document.name = "UI_DialoguePannel";
            contentLabel = Container.Q<Label>("content");
            speakerLabel = Container.Q<Label>("speaker");
            playerActions.DialoguePanel.Continue.started += (ctx) =>
            {
                states.currentLine++;
                if (states.currentLine >= props.lines.Length)
                {
                    UIStack.Pop();
                }
                else
                {
                    states.displayedText = string.Empty;

                }
                states.currentLine %= props.lines.Length;
            };
            timer.OnTick.AddListener(MarkDirty);
            return Task.CompletedTask;
        }


        protected override void LoadStatesFromProps(Props props, out States states)
        {
            states = new States()
            {
                displayedText = string.Empty,
                currentLine = 0
            };
            // 已经启动，表示现在的Props 
            timer.Clear();
            timer.Set(c_PrintInterval);

        }

        public override void OnActive()
        {
            //StartPrintCounter();
            timer.Set(c_PrintInterval);
            playerActions.DialoguePanel.Continue.Enable();
            PlayerUtil.FrozeCurrentCharacter();
            MiscUtil.ShowCursor();
        }


        public override void OnDeactive()
        {
            //StopCurrentPrintCounterIfAny();
            timer.Clear();
            playerActions.DialoguePanel.Continue.Disable();
            PlayerUtil.UnfrozeCurrentCharacter();
            MiscUtil.HideCursor();
        }

        /// <summary>
        /// 实现打字机效果
        /// </summary>
        public override void Render()
        {
            speakerLabel.text = base.props.speaker;
            if (states.displayedText.Length < CurrentLineLength())
            {
                states.displayedText = CurrentLine()[..(states.displayedText.Length + 1)];
                contentLabel.text = states.displayedText;
            }

            
            if (states.displayedText.Length == CurrentLineLength())
            {
                OnDisplayFullCotent.Invoke();
                if (states.currentLine >= props.lines.Length - 1)
                {
                    timer.Clear();
                }
            }
        }

        private string CurrentLine()
        {
            return base.props.lines[states.currentLine];
        }

        private int CurrentLineLength()
        {
            return base.props.lines[states.currentLine].Length;
        }

        #region [[Inherited Fields]]
        public override string Address => "PUI_DialoguePannel";
        protected override bool NeedPreload => true;
        public override bool IsResident => true;
        public override bool BlockPlayerControl => true;
        public override Props DefaultProps { get; }
        #endregion
    }
}

File: src/Assets/LuaScript/UI/dialog_panel.lua
local dinames = require 'dinames'
local util = require 'util'
--[[ 
Props {
    speaker => string,
    lines => string[],
    options => string[] -- 请务必按原始类型传到Lua
}
]]-- 


address= 'PUI_DialoguePannel'
is_resident = true
need_preload = true

local PRINT_INTERVAL = 0.15

local lb_content = nil
local ve_option_list = nil

local ui_option_list = nil
local player_actions = nil
local uistack = nil
local timer = nil



function loadStatesFromProps(props)
    local new_states = CS.QS.Common.TypelessDict()
    new_states:Set("text_length", 1)
    new_states:Set("current_line", 0)

    timer:Clear()
    if self.IsVisible then
        timer:Set(PRINT_INTERVAL);
    end

    -- 数据自顶向下流动
     oplist_props = CS.QS.Common.TypelessDict()
    oplist_props:Set("options", props:Get("options")[0])
    ui_option_list.props = oplist_props

    return new_states
end

function onDocumentLoaded()
    local source = TaskCompleteSource()


    self.Document.name = 'UI_DialogPannel'
    lb_content = uidoc.uquery.Q(self.Container, "content")
    ve_option_list = uidoc.uquery.Q(self.Container, "optionList")

    player_actions = Trunk.Context:GetInstance(dinames.PlayerControls)
    timer = Trunk.Context:GetInstance(dinames.Timer)
    uistack = Trunk.Context:GetInstance(dinames.UIStack)
    DebugLog(player_actions)


    player_actions.DialoguePanel.Continue:started('+', function (_)
            local new_line_index = self.states:Get("current_line") + 1
            
            if(new_line_index >= self.props:Get("lines").Length) then
                uistack:Pop()
            else
                -- 想办法让这也响应式上
                local  oplist_props = CS.QS.Common.TypelessDict()
                oplist_props:Set("options", self.props:Get("options")[new_line_index])
                ui_option_list.props = oplist_props
                
                self.states:Set("current_line", new_line_index)
                self.states:Set("text_length", 0)
            end
        end)

    timer.OnTick:AddListener(function ()
        local new_len = self.states:Get("text_length") + 1
        if new_len <= LuaUtil.strlen(self.props:Get("lines")[self.states:Get("current_line")]) then
            self.states:Set("text_length", new_len)
        end
        self:MarkDirty()
    end)

    local load_option_list_ui = LuaUIDocumentLoadOp(ve_option_list, "Lua_UI_DialogPanel_OptionList", self)
    load_option_list_ui:Invoke()
    util.await(load_option_list_ui.Task, function (result)
        ui_option_list = result
        -- DebugLog(ui_option_list)
        source:SetResult(0)
    end)



    return source.Task
end


function onActive()
    timer:Set(PRINT_INTERVAL)
    player_actions.DialoguePanel.Continue:Enable()
    PlayerUtil.FrozeCurrentCharacter()
end


function onDeactive()
    timer:Clear()
    player_actions.DialoguePanel.Continue:Disable()
    PlayerUtil.UnfrozeCurrentCharacter()
end


function render()
    local line = self.props:Get("lines")[self.states:Get("current_line")]
    local length = self.states:Get("text_length")
    lb_content.text = LuaUtil.substr(line, 1, length)
    
end